### PR TITLE
Add testnet bonus page

### PR DESCRIPTION
**Pull request by thunderbiscuit**
## What
This draft PR adds a _Raspibolt on Testnet_ bonus guide page.

This PR is currently in draft mode because I'm still working on the Lightning part of the RaspiBolt guide, and so I expect more sections will be added before I'm through.

[See the in-progress page here](https://thunderbiscuit.github.io/raspibolt/bonus/bitcoin/testnet.html).

## Why
Testnet is awesome, and to quote the page: _Helping bitcoiners run a full testnet setup is a goal worthy of the RaspiBolt._

## Scope
- [x] independent bonus guide

### CODE REVIEW

**Comment by Stadicus** [comment]
What do you think about running testnet in parallel to mainnet? I think that'd be trivial, because applications can be started multiple times with additional systemd services, different base directories and config files. It would be more reliable (no copying config files back and forth), and more flexible.

We just need to be careful that every configuration runs with separate ports. Mainnet and testnet would just be activated / deactivated by running or disabling the related systemd services.

We could create dedicated directories for this:

### Bitcoin

* `/data/bitcoin-testnet/` --> contains the testnet `bitcoin.conf` and all data in the `testnet3` subdirectory
* `/data/electrs-testnet/` --> contains the testnet `electrs.config` and testnet data
* `/home/btc-rpc-explorer/` --> not quite sure yet if the whole application needs to be duplicated, or if a separate `.env` testnet config file would be doable
* separate Tor hidden services can be added
* UFW opens additional port 60002 for Electrs testnet

### LND

* `/data/lnd-testnet` --> contains `lnd.conf`, separate wallet and data 
* `/home/rtl/RTL` --> not quite sure yet if the whole application needs to be duplicated, or if a separate `RTL-Config.json` testnet config file would be doable
* separate Tor hidden services can be added

WDYT?

**Comment by VajraOfIndra** [comment]
FYI, there is an outdated bonus guide by Rob Clark that had the same aim: [https://github.com/robclark56/RaspiBolt-Extras/blob/master/RB_extra_03.md](https://github.com/robclark56/RaspiBolt-Extras/blob/master/RB_extra_03.md). 

It has a beautiful diagram that explains how mainnet and testnet communication would work with bitcoind and lnd. It would be really nice to reuse it and maybe update it with Electrs, Explorer ect:
diagram

> What do you think about running testnet in parallel to mainnet? 

The question is address at @thunderbiscuit but if you allow me to chime in in the meantime... I love the idea! :100: :)

**Comment by thunderbiscuit** [comment]
Hey thanks for checking out the PR so quickly! 

I do think there is a ton of demand for testnet setups, and we should try and provide information for different use cases. I wrote a blog post about running testnet on Umbrel over a year ago, and I get a lot of action on it. People DM me about it, open issues on the Gist, etc. Unfortunately Umbrel really isn't built for that. It's evolved a lot since (and is always rapidly iterating), and I now think trying to fit testnet on it is just not that great of an idea. But I can attest there is a lot of demand for that sort of thing.

I think there is value in having both pages. Here are some of my thoughts on this:

## All networks on the same setup
Running both mainnet and testnet at the same time I assume must be very popular, because I've seen many tutorials/guides and questions on it on all the main node implementations. It's a bit more to manage and build, and so I think it deserves a page for sure. A few things come to mind:
1. It does sound like it implies quite a few differences from the core RaspiBolt guide, so that would make me think the page might be quite advanced, as it needs to inject more fixes into the main "path" for raspibolt than just a few new config files.
2. I totally agree with @VajraOfIndra, the diagram is great. One thing that wasn't clear for me also was how Tor played in all of that, so if we can find a way to incorporate that in the diagram I think it would be a plus.

## Testnet-only setup
To explain why I wanted a testnet page (and just a testnet node for myself in general), I'll expand a bit more about my use case for it. I've never had a full mainnet and testnet setup on the same machine; I find them too different in scope and usage (or at least my use of it is). For example, in this case for my raspibolt testnet node I was able to buy a $10 a month cloud instance with 50GB storage and 2GB memory which I expect to keep only for a month or two (maybe more if it's just too much fun!). The testnet setup is fantastic; (a) it's cheap, (b) I don't care if I mess up and I can really try out anything I want (I'm on my third complete wipe of the cloud instance in 10 days, and it's gloriously freeing). The fact that I _don't_ have mainnet on the same machine is the very thing that makes it valuable to me. I don't worry about starting the wrong bitcoin daemon, and didn't have to buy 1TB of storage.

For those reasons, the idea of running both on the same machine doesn't quite fit my use case, and the complexity of the guide for running both was a bit scary (the guide page from Rob Clark I mean), particularly because I was trying to extract just the testnet parts from it. Also, I once broke a Raspiblitz node because I tried running testnet on it after having my real, mainnet node up, and messed up the configurations because I didn't know enough about it. Bitcoind started downloading the testnet blocks right on the microSD card with the OS instead of the SSD external drive, filled it up in no time, and I wasn't able to get back in to fix the node to fix after the OS crashed. Had to wipe the whole thing (or I wasn't savvy enough to recover anyway). #reckless I guess? 😆 It turned me off of "playing around" with nodes for a long time, which I now know is really the key to learning. I think optimally, most tinkerers should have a sort of bi-modal approach to nodes: one setup that is super safe, sacred, that you're very very careful with, and one you're totally carefree about, 100% reckless, don't-care-if-I-break-it type. The type of node you'll install LND _and_ C-lightning _and_ Eclair on, just because it's fun and you want to try them out. Those two type of nodes being on different hardware is key to unleashing their respective power in my opinion. For that reason, I also think a testnet-only page is important.

tldr; LET'S DO BOTH

**Comment by Stadicus** [comment]
Lol, best tl;dr I've seen in a long time 😁

**Comment by Stadicus** [comment]
Your reasoning makes total sense. I think a simple bonus guide that contains all the alternative configuration files makes total sense. 

We can build an additional guide for a testnet configuration running in parallel to mainnet from there later.

**Comment by VajraOfIndra** [comment]
> Your reasoning makes total sense. I think a simple bonus guide that contains all the alternative configuration files makes total sense.
> 
> We can build an additional guide for a testnet configuration running in parallel to mainnet from there later.

So if I understood well, there will be two new bonus guides in the end:

1. A bonus guide to build a bitcoin and LN node on testnet (only)
2. A bonus guide to run testnet alongside mainnet on the same node.

And the main guide will not have testnet included, correct? 

Sounds like a safe way to do it.

**Comment by RooSoft** [comment]
>

**Comment by RooSoft** [comment]
I was about to install testnet on a regular raspibolt, so having both on the same machine made sense at first glance.

Thing is, my mainnet raspibolt does not have any channel yet. So it pretty much is a dev machine and thus not a very frequent use case.

@thunderbiscuit made me realize it doesn't make much sense to run both mainnet and testnet on the same machine when the mainnet is being used for real. Testnet nodes value is partly derived from the fact that they can be reset at will.

On another note, buying all that hardware for a test node seems a bit overkill... so why bother documenting it at all if it's not a typical use case?

Still, documenting installations such as thunderhub on a raspibolt which is active might not be the best idea. So I am on the fence about this... I think build the docs based on a real physical raspibolt might end up providing the most value.

**Comment by RooSoft** [comment]
For the time being, I really need a testnet raspibolt, so I will configure one and document it along the way on a separate wiki...

If we end up considering it's a good addition to Raspibolt, I will be glad to share it. We have to consider if might be worth spending time keeping that doc up to date in the meantime.

**Comment by RooSoft** [comment]
Sorry about the PR close, that was a fat finger error...

**Comment by thunderbiscuit** [comment]
I think this page is ready for review. It keeps things very simple, not trying to rewrite the guide, but simply provide small fixes for the different configurations and commands required to enable the setup for testnet rather than mainnet. 

If you'd like to see anything else added or removed, let me know!

File: bonus/bitcoin/testnet.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+nav_exclude: true
+has_children: false
+has_toc: false
+---
+
+# Bonus guide: Raspibolt on Testnet
+{: .no_toc }
+
+Difficulty: Intermediate
+{: .label .label-yellow }
```
Comment (VajraOfIndra) [inline bonus/bitcoin/testnet.md:15]:
```suggestion
Difficulty: Medium
{: .label .label-yellow }

Status: Tested v3
{: .label .label-green }
```

Adding the Tested label and changed the difficulty to Medium (now the preferred term over Intermediate)

File: bonus/bitcoin/testnet.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+<details open markdown="block">
+  <summary>
+    Table of contents
+  </summary>
+  {: .text-delta }
+1. TOC
+{:toc}
+</details>
```
Comment (VajraOfIndra) [inline bonus/bitcoin/testnet.md:24]:
```suggestion
---

Table of contents
{: .text-delta }

1. TOC
{:toc}

---
```

This is the new format we use for the TOC

File: bonus/bitcoin/testnet.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+## Introduction
+Running a testnet node is a great way to get acquainted with the Raspibolt and the suite of Bitcoin-related software typical of these powerful setups. Moreover, testnet empowers users to tinker with the software and its many configurations without the threat of losing funds. Helping bitcoiners run a full testnet setup is a goal worthy of the RaspiBolt, and this page should provide you with the knowledge to get there.
```
Comment (VajraOfIndra) [inline bonus/bitcoin/testnet.md:27]:
```suggestion
Running a testnet node is a great way to get acquainted with the RaspiBolt and the suite of Bitcoin-related software typical of these powerful setups. Moreover, testnet empowers users to tinker with the software and its many configurations without the threat of losing funds. Helping bitcoiners run a full testnet setup is a goal worthy of the RaspiBolt, and this page should provide you with the knowledge to get there.
```
Fixed small typo

File: bonus/bitcoin/testnet.md [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+The great news is that most of the RaspiBolt guide can be used as-is. The small adjustments come in the form of changes to the config files and ports for testnet. You can follow the guide and simply replace the following configurations in the right places as you go.
+
+## Bitcoin daemon
+File location: `/data/bitcoin/bitcoin.conf`
+```conf
```
Comment (VajraOfIndra) [inline bonus/bitcoin/testnet.md:33]:
```suggestion
```ini
```

I think we usually use `ini` for the configuration scripts language code.. but tbh, `conf` might be a better choice, I don't know enough about these conventions. So, I'm just flaggign it to start a discussion and see what is the best option here :)

**Review by VajraOfIndra — CHANGES_REQUESTED** [review]
Just a short review to highlight a few changes to align with the formatting of the rest of the guide :)

I didn't test these settings and never used testnet so I'll leave the review and approval to someone else who has a testnet node running (@RooSoft if I recall well?)

**Comment by RooSoft** [comment]
> I didn't test these settings and never used testnet so I'll leave the review and approval to someone else who has a testnet node running (@RooSoft if I recall well?)

Yeah, I made one and took notes along the way....

Once conflicts gets solved on that PR, and requested changes made, I will review it.

File: bonus/bitcoin/testnet.md (hunk 3 above)
Comment (Stadicus) [reply bonus/bitcoin/testnet.md:27]: Great introduction, I love it!

File: bonus/bitcoin/testnet.md [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+bind=127.0.0.1
+rpcbind=127.0.0.1
+
+[signet]
+# rpcport=38332
+
+[regtest]
+# rpcport=18443
+```
+
```
Comment (Stadicus) [inline bonus/bitcoin/testnet.md:87]: We could add some additional information and actionable hints as well:

* The testnet data is stored in `/data/bitcoin/testnet3`
* To check the log file, use `tail -f /data/bitcoin/testnet3/debug.log`

Comment (Stadicus) [inline bonus/bitcoin/testnet.md:87]: As the authentication cookie is located in `/data/bitcoin/testnet3/.cookie`, the command in the systemd unit "bitcoind.service" to grant the group "bitcoin" read permissions is no longer correct.

This means that `bitcoin-cli` (and all other applications) can't access the bitcoins rpc api.

The following line in `/etc/systemd/system/bitcoind.service` needs to be changed:
```
-startupnotify="chmod g+r /home/bitcoin/.bitcoin/testnet3/.cookie"
```

File: bonus/bitcoin/testnet.md (hunk 1 above)
Comment (thunderbiscuit) [reply bonus/bitcoin/testnet.md:15]: 👍

File: bonus/bitcoin/testnet.md [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+## Tor
+File location: `/etc/tor/torrc`
+```shell
+############### This section is just for location-hidden services ###
+HiddenServiceDir /var/lib/tor/hidden_service_electrs/
+HiddenServiceVersion 3
+HiddenServicePort 60002 127.0.0.1:60002
+```
+
```
Comment (Stadicus) [inline bonus/bitcoin/testnet.md:132]: Maybe helpful

```
$ sudo systemctl reload tor
$ sudo cat /var/lib/tor/hidden_service_electrs/hostname
```

File: bonus/bitcoin/testnet.md (hunk 2 above)
Comment (thunderbiscuit) [reply bonus/bitcoin/testnet.md:24]: 👍

File: bonus/bitcoin/testnet.md [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+# Electrs settings
+electrum_rpc_addr = "127.0.0.1:60001"
+db_dir = "/data/electrs/db/"
+index_lookup_limit = 1000
+
+# Logging
+log_filters = "INFO"
+timestamp = true
+```
+
```
Comment (Stadicus) [inline bonus/bitcoin/testnet.md:110]: Not strictly necessary, but maybe helpful? Everyone will want to check the log to see if it works... :)

```
$ sudo journalctl -u electrs -f
```

File: bonus/bitcoin/testnet.md [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+upstream electrs {
+  server 127.0.0.1:60001;
+}
+
+server {
+  listen 60002 ssl;
+  proxy_pass electrs;
+}
+```
+
```
Comment (Stadicus) [inline bonus/bitcoin/testnet.md:123]: Could be helpful:

```
$ sudo nginx -t
$ sudo systemctl reload nginx
$ sudo ufw allow 60002/tcp comment 'allow Electrum SSL Testnet'
```

File: bonus/bitcoin/testnet.md (hunk 4 above)
Comment (thunderbiscuit) [reply bonus/bitcoin/testnet.md:33]: Since you're using `ini` in other places, I made the switch. I'm not as familiar with that category/language of syntax highlighting, but it does seem to work very well for those files.

File: bonus/bitcoin/testnet.md [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+## Tor
+File location: `/etc/tor/torrc`
+```shell
+############### This section is just for location-hidden services ###
+HiddenServiceDir /var/lib/tor/hidden_service_electrs/
+HiddenServiceVersion 3
+HiddenServicePort 60002 127.0.0.1:60002
+```
+
+## Interacting with the LND daemon
```
Comment (Stadicus) [inline bonus/bitcoin/testnet.md:133]: In `/data/lnd/lnd.conf`, the following change is necessary for LND to run:

```
bitcoin.active=1
#bitcoin.mainnet=1
```

File: bonus/bitcoin/testnet.md (hunk 3 above)
Comment (thunderbiscuit) [reply bonus/bitcoin/testnet.md]: Thanks for the typo. Found another one too!

File: bonus/bitcoin/testnet.md (hunk 9 above)
Comment (Stadicus) [reply bonus/bitcoin/testnet.md:133]: The following command gives members of group "lnd" permission to traverse the LND directories to reach the macaroons.

```
$ sudo chmod -R g+X /data/lnd/data/
```

**Review by Stadicus — COMMENTED** [review]
I started going through this valuable bonus guide and left a few general comments for now.

File: bonus/bitcoin/testnet.md (hunk 5 above)
Comment (thunderbiscuit) [reply bonus/bitcoin/testnet.md]: Good catch!

Comment (thunderbiscuit) [reply bonus/bitcoin/testnet.md]: 👍

File: bonus/bitcoin/testnet.md (hunk 6 above)
Comment (thunderbiscuit) [reply bonus/bitcoin/testnet.md]: 👍

File: bonus/bitcoin/testnet.md (hunk 7 above)
Comment (thunderbiscuit) [reply bonus/bitcoin/testnet.md]: 👍

File: bonus/bitcoin/testnet.md (hunk 8 above)
Comment (thunderbiscuit) [reply bonus/bitcoin/testnet.md]: 👍

File: bonus/bitcoin/testnet.md (hunk 9 above)
Comment (thunderbiscuit) [reply bonus/bitcoin/testnet.md]: 👍

**Comment by thunderbiscuit** [comment]
Thanks for the review Stadicus! Now I'm not sure how I got my stuff working because in some cases I don't remember doing those... but it was a few weeks back so maybe I just don't remember. In any case, I think I got all the added stuff in.

**Review by VajraOfIndra — APPROVED** [review]

File: guide/bonus/bitcoin/testnet.md [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+HiddenServicePort 60002 127.0.0.1:60002
+```
+
+Once that's done, you'll need to start the service using:
+```sh
+$ sudo systemctl reload tor
+$ sudo cat /var/lib/tor/hidden_service_electrs_testnet/hostname
+```
+
+## LDN
```
Comment (VajraOfIndra) [inline guide/bonus/bitcoin/testnet.md:160]:
```suggestion
## LND
```

File: guide/bonus/bitcoin/testnet.md [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+log_filters = "INFO"
+timestamp = true
+```
+
+Check the logs using
+```sh
+$ sudo journalctl -u electrs -f
+```
+
+## NGINX
```
Comment (VajraOfIndra) [inline guide/bonus/bitcoin/testnet.md:123]:
```suggestion
### NGINX
```
I propose to make this a third-level heading as it is still part of the Electrs section.

File: guide/bonus/bitcoin/testnet.md [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+```
+
+Your nginx might need to be reloaded using:
+```sh
+$ sudo nginx -t
+$ sudo systemctl reload nginx
+$ sudo ufw allow 60002/tcp comment 'allow Electrum SSL Testnet'
+```
+
+## Tor
```
Comment (VajraOfIndra) [inline guide/bonus/bitcoin/testnet.md:143]:
```suggestion
### Tor
```

Same as above, I propose to make this a third-level heading as it is still part of the Electrs section.

**Review by VajraOfIndra — APPROVED** [review]
utACK, just a couple of suggestions regarding the structure.

@RooSoft @Stadicus shall we merge this guide or would you prefer to make some testing?

**Comment by kristapsk** [comment]
Maybe it's good idea to do also signet guide? Signet is nowadays often more practical for most test scenarious. Haven't checked that all involved sofware supports signet in addition to testnet3 though.

File: guide/bonus/bitcoin/testnet.md (hunk 10 above)
Comment (thunderbiscuit) [reply guide/bonus/bitcoin/testnet.md:160]: Ah yep good catch!

File: guide/bonus/bitcoin/index.md [hunk 13]
```diff
@@ -37,4 +37,10 @@ has_toc: false
 
 ---
 
+## Testnet
+
+* **[Running a Testnet Raspibolt Node](testnet.md)** - configure your RaspiBolt for use on testnet
```
Comment (Stadicus) [inline guide/bonus/bitcoin/index.md:42]: s/Raspibolt/RaspiBolt

File: guide/bonus/bitcoin/testnet.md [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+has_children: false
+has_toc: false
+---
+
+# Bonus guide: RaspiBolt on Testnet
+{: .no_toc }
+
+Difficulty: Medium
+{: .label .label-yellow }
+
```
Comment (Stadicus) [inline guide/bonus/bitcoin/testnet.md:16]: Adding a quick TL;DR would be good:

* we made it standard across all updated bonus guides
* it helps align the color badges

Maybe something like

> You can run your RaspiBolt node on testnet to develop and experiment with new applications, without putting real money at risk. This bonus guide highlights all configuration changes compared to the main guide.

File: guide/bonus/bitcoin/testnet.md [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+Your nginx might need to be reloaded using:
+```sh
+$ sudo nginx -t
+$ sudo systemctl reload nginx
+$ sudo ufw allow 60002/tcp comment 'allow Electrum SSL Testnet'
+```
+
+### Tor
+Create a separate service for testnet over Tor:
```
Comment (Stadicus) [inline guide/bonus/bitcoin/testnet.md:144]: Maybe 

> Create a separate service for testnet over Tor by adding the following lines in the `location-hidden services` section:

File: guide/bonus/bitcoin/testnet.md [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+HiddenServicePort 60002 127.0.0.1:60002
+```
+
+Once that's done, you'll need to start the service using:
+```sh
+$ sudo systemctl reload tor
+$ sudo cat /var/lib/tor/hidden_service_electrs_testnet/hostname
+```
+
+## LND
```
Comment (Stadicus) [inline guide/bonus/bitcoin/testnet.md:160]: This is a bit short, as it's not the full configuration. I'd at least mention that the configuration below is only the diff, and probably list the whole `[bitcoin]` section:

> Change the following lines in the `[bitcoin]` section:

```ini
[Bitcoin]
bitcoin.active=1
#bitcoin.mainnet=1
bitcoin.testnet=1
bitcoin.node=bitcoind
```

**Review by Stadicus — COMMENTED** [review]
Finally, I took the review of this pull request up again. I think a few minor last changes would help, but that should be it.

Two improvements for future iterations could be:
* harmonize the config files for mainnet / testnet (so they're not completely different, like for Bitcoin Core)
* add a section on where to get Testnet Bitcoins

But I don't want to overburden this change, I'm happy if it's in!

**Comment by thunderbiscuit** [comment]
Thanks for the input! I applied all fixes. Cheers!
